### PR TITLE
Add ECMWF Metview recipe

### PR DESCRIPTION
**Pull request by dtip**

### CODE REVIEW

**Comment by dtip** [comment]
@StephanSiemen I've added you as a maintainer. Please confirm that it's okay with you!

**Comment by StephanSiemen** [comment]
> @StephanSiemen I've added you as a maintainer. Please confirm that it's okay with you!

Hi @dtip, thanks for adding me as a maintainer. Happy to help look after this package.

File: recipes/metview/build.sh [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+    echo ERROR: An error was encountered with the build.
+    dump_output
+    exit 1
+}
+
+# If an error occurs, run our error handler to output a tail of the build.
+trap 'error_handler' ERR
+
+# Set up a repeating loop to send some output to Travis.
+bash -c "while true; do echo \$(date) - building ...; sleep $PING_SLEEP; done" &
```
Comment (tadeu) [inline recipes/metview/build.sh]: Looks like this is useful for "debugging" a build, but consider removing it from the final recipe

File: recipes/metview/build.sh [hunk 2]
```diff
@@ -0,0 +1,57 @@
+#!/usr/bin/env bash
+
+set -e # Abort on error.
+
+export PING_SLEEP=30s
+export WORKDIR="$( cd "$( dirname "${BASH_SOURCE[0]}" )" && pwd )"
```
Comment (tadeu) [inline recipes/metview/build.sh]: It's more clear to give preference to variables from `conda build`: https://docs.conda.io/projects/conda-build/en/latest/user-guide/environment-variables.html (`RECIPE_DIR` in this case)

File: recipes/metview/build.sh [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+bash -c "while true; do echo \$(date) - building ...; sleep $PING_SLEEP; done" &
+PING_LOOP_PID=$!
+
+export PYTHON=
+export LDFLAGS="$LDFLAGS -L$PREFIX/lib -Wl,-rpath,$PREFIX/lib"
+export CFLAGS="$CFLAGS -fPIC -I$PREFIX/include"
+
+SRC_DIR="$(pwd)"
+BUILD_DIR=$SRC_DIR/../build
+mkdir $BUILD_DIR
```
Comment (tadeu) [inline recipes/metview/build.sh]: `mkdir build && cd build` is simpler

File: recipes/metview/build.sh [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+cd $BUILD_DIR
+
+export TMPDIR=/tmp/
```
Comment (tadeu) [inline recipes/metview/build.sh]: Is this really needed? It's assuming something from the build machine

File: recipes/metview/build.sh [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+
+if [[ $(uname) == Linux ]]; then
+    export PATH=$PATH:$BUILD_PREFIX/x86_64-conda_cos6-linux-gnu/sysroot/usr/
```
Comment (tadeu) [inline recipes/metview/build.sh]: Why is this needed?

**Comment by dtip** [comment]
@tadeu thanks for the comments!

File: recipes/metview/build.sh (hunk 1 above)
Comment (dtip) [reply recipes/metview/build.sh]: No, the point is to limit the amount of log output. We ran into some trouble a while back (2017 according to git history) because we were hitting the Travis log line limit when building some ECMWF packages. Perhaps something has changed since then - if you can confirm it's not needed then I'll delete it.

File: recipes/metview/build.sh (hunk 2 above)
Comment (dtip) [reply recipes/metview/build.sh]: Done

File: recipes/metview/build.sh (hunk 3 above)
Comment (dtip) [reply recipes/metview/build.sh]: Agreed

File: recipes/metview/build.sh (hunk 5 above)
Comment (dtip) [reply recipes/metview/build.sh]: Thought I commented here. I'm having trouble getting cmake to find libpthread.so. I tried setting `CMAKE_LIBRARY_PATH= $BUILD_PREFIX/x86_64-conda_cos6-linux-gnu/sysroot/usr/lib` which worked (as in, cmake found pthread) but the travis builds started started failing at compile time with the error 'cannot find any c preprocessor' which is a bit bizarre.

It seems that using $LDFLAGS isn't working either - cmake still can't find pthread. I'll keep chipping away at it until it works.

**Comment by dtip** [comment]
@tadue I'm seeing the error
```
[  0%] Generating rpcmars.h
cannot find any C preprocessor (cpp)
rpcgen: C preprocessor failed with exit code 1
make[2]: *** [mars-client/src/rpcmars.h] Error 1
make[1]: *** [mars-client/src/CMakeFiles/marsxdr.dir/all] Error 2
make[1]: *** Waiting for unfinished jobs....
```
Google isn't turning up much. Any ideas?

**Comment by tadeu** [comment]
> I'm seeing the error
> ```
> [  0%] Generating rpcmars.h
> cannot find any C preprocessor (cpp)
> rpcgen: C preprocessor failed with exit code 1
> make[2]: *** [mars-client/src/rpcmars.h] Error 1
> make[1]: *** [mars-client/src/CMakeFiles/marsxdr.dir/all] Error 2
> make[1]: *** Waiting for unfinished jobs....```

It's because indeed there is no `cpp` in the PATH in `conda build`, and there is probably something hardcoded in the build scripts trying to look for `cpp`. The compiler packages define environment variables such as `$CPP` that point to a place like `$PREFIX/bin/x86_64-conda_cos6-linux-gnu-cpp`.

CMake will usually work out-of-the-box with these environment variable, but hardcoded things should be changed to respect the environment variables (they are standard).

File: recipes/metview/meta.yaml [hunk 6]
```diff
@@ -0,0 +1,68 @@
+{% set version = "5.6.0" %}
+
+package:
+  name: metview
+  version: {{ version }}
+
+source:
+  url: https://confluence.ecmwf.int/download/attachments/30048389/Metview-{{ version }}-dev3-Source.tar.gz
```
Comment (tadeu) [inline recipes/metview/meta.yaml]: An attachment in confluence seems hard to update, specially because of that id `30048389` part. It would be better if it was a link to PyPI or to a GitHub or Bitbucket tagged release, for example.

Comment (dtip) [reply recipes/metview/meta.yaml]: I believe the id number is a page id - it's constant as long as the tarballs are all attached to the same page on confluence (which they are at ECMWF). Certainly the other ECMWF feedstocks have been using confluence like this without issue!

File: recipes/metview/build.sh (hunk 1 above)
Comment (dtip) [reply recipes/metview/build.sh]: Ah but we no longer use Travis on conda-forge do we! This work around is not necessary on Azure - I'll remove it.

**Comment by dtip** [comment]
I'm going to have to leave this sitting around for a few days while I'm traveling. It's not forgotten - I'll get back to it soon!

**Comment by dtip** [comment]
Code is compiling at last! Currently 7/472 tests are failing so it still needs a little more work.

**Comment by dtip** [comment]
@tadeu The build is hitting the max time limit on Travis. CircleCI and Azure and fine. What can I do about it?

**Comment by mariusvniekerk** [comment]
@dtip, its okay if it times out for travis, we don't build on travisci anymore by default for osx

**Comment by dtip** [comment]
Thanks for the info @mariusvniekerk . I think I'll add the osx build at a later date - there are some failing tests which need a bit of investigation.

@conda-forge/help-c-cpp could you please review?

File: recipes/metview/meta.yaml [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+  host:
+    - curl
+    - eccodes
+    - magics
+    - boost
+    - qt
+    - flex
+    - bison
+    - libnetcdf
+    - netcdf-cxx4=4.2
```
Comment (chrisburr) [inline recipes/metview/meta.yaml]: There needs to be a space here:
```suggestion
    - netcdf-cxx4 =4.2
```

File: recipes/metview/meta.yaml [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+    - fftw
+  run:
+    - eccodes
+    - magics
+    - boost
+    - qt
+    - flex
+    - bison
+    - libnetcdf
+    - netcdf-cxx4=4.2
```
Comment (chrisburr) [inline recipes/metview/meta.yaml]:
```suggestion
    - netcdf-cxx4 =4.2
```

File: recipes/metview/meta.yaml (hunk 7 above)
Comment (dtip) [reply recipes/metview/meta.yaml]: @chrisburr thanks for fixing & merging!

Could you explain why a space is needed? Why did the automatic linter not catch the above code if it's a problem?

Comment (chrisburr) [reply recipes/metview/meta.yaml]: So far as I know the specification requires there to be a space but it's only enforced in some parts of `conda-build`. Creating feedstocks after recipes are merged in `staged-recipes` often fails if it's not included.

It's something that should be added to the linter at some point.

Comment (dtip) [reply recipes/metview/meta.yaml]: Thanks for the info :)
